tests/test_providers.py

```diff
@@ -312,3 +324,632 @@ def parse_(self, response: DummyResponse, page: GeoProductPage):  # type: ignore
+
+    assert len(params) == 1
+    assert params[0] == {
+        "url": url,
+        "product": True,
+        "browserHtml": True,
+        "productOptions": product_options,
+    }
+
+    assert type(item["page"].response) == AnyResponse
+    assert type(item["page"].response.response) == BrowserResponse
+    assert type(item["page"].browser_response) == BrowserResponse
+    assert type(item["page"].product) == Product
+
+    assert id(item["page"].browser_response) == id(item["page"].response.response)
+
+
+@ensureDeferred
+async def test_provider_any_response_product_extract_from_http_response(mockserver):
+    @attrs.define
+    class SomePage(BasePage):
+        response: AnyResponse
+        product: Product
+
+    class ZyteAPISpider(Spider):
+        def start_requests(self):
+            yield Request(self.url, callback=self.parse_)
+
+        def parse_(self, response: DummyResponse, page: SomePage):
+            yield {"page": page}
+
+    product_options = {"extractFrom": "httpResponseBody"}
+    settings = provider_settings(mockserver)
+    settings["ZYTE_API_PROVIDER_PARAMS"] = {"productOptions": product_options}
+    item, url, crawler = await crawl_single_item(ZyteAPISpider, HtmlResource, settings)
+    params = crawler.engine.downloader.handlers._handlers["http"].params
+
+    assert len(params) == 1
+    assert params[0] == {
+        "url": url,
+        "product": True,
+        "httpResponseBody": True,
+        "productOptions": product_options,
+        "httpResponseHeaders": True,
+    }
+
+    assert type(item["page"].response) == AnyResponse
+    assert type(item["page"].response.response) == HttpResponse
+    assert type(item["page"].product) == Product
+
+
+@ensureDeferred
+async def test_provider_any_response_product_options_empty(mockserver):
+    @attrs.define
+    class SomePage(BasePage):
+        response: AnyResponse
+        product: Product
+
+    class ZyteAPISpider(Spider):
+        def start_requests(self):
+            yield Request(self.url, callback=self.parse_)
+
+        def parse_(self, response: DummyResponse, page: SomePage):
+            yield {"page": page}
+
+    settings = provider_settings(mockserver)
+    settings["ZYTE_API_PROVIDER_PARAMS"] = {"productOptions": {}}
+    item, url, crawler = await crawl_single_item(ZyteAPISpider, HtmlResource, settings)
+    params = crawler.engine.downloader.handlers._handlers["http"].params
+
+    assert len(params) == 1
+    assert params[0] == {
+        "url": url,
+        "product": True,
+        "browserHtml": True,
+    }
+
+    assert type(item["page"].response) == AnyResponse
+    assert type(item["page"].response.response) == BrowserResponse
+    assert type(item["page"].product) == Product
+
+
+# The issue here is that HttpResponseProvider runs earlier than ScrapyZyteAPI.
+# HttpResponseProvider doesn't know that it should not run since ScrapyZyteAPI
+# could provide HttpResponse in anycase.
```

Comment:
What is the failure? Sorry, I haven't checked the logs :)

Reply:
The issue happens if `HttpResponse` is explicitly declared as a dependency, say in the PO.

Since `HttpResponseProvider` runs much earlier than `ZyteApiProvider`, it would make a request to ZAPI. When it's `ZyteApiProvider` turn to fullfill dependencies, it would have a 2nd request to ZAPI to fulfill the `AnyResponse` + `Product` dependencies.

The ideal scenario would be having only a single ZAPI request that would fulfill all three of the `HttpResponse`, `AnyResponse`, and `Product` dependencies.

It could be the case that in another PR, we can combine `HttpResponseProvider` and `ZyteApiProvider` together, or perhaps create a decision mechanism to determine which providers should run that would result in a more optimal dependency creation.

I'm not sure how often this would occur in practice though, since if you have `AnyResponse`, there's not much need to declare an `HttpResponse` dependency. With that, it should be easy to avoid.